ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java

```diff
@@ -0,0 +1,1021 @@
+    int index = ((RexInputRef) expr).getIndex();
+    return rowResolver.getColumnInfos().get(index).getInternalName();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected boolean isExprsListExpr(Object o) {
+    return o instanceof HiveRexExprList;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected List<RexNode> getExprChildren(RexNode expr) {
+    if (expr instanceof RexCall) {
+      return ((RexCall) expr).getOperands();
+    } else if (expr instanceof HiveRexExprList) {
+      return ((HiveRexExprList) expr).getExpressions();
+    }
+    return new ArrayList<>();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected TypeInfo getTypeInfo(RexNode expr) {
+    return expr.isA(SqlKind.LITERAL) ?
+        TypeConverter.convertLiteralType((RexLiteral) expr) :
+        TypeConverter.convert(expr.getType());
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected List<TypeInfo> getStructTypeInfoList(RexNode expr) {
+    StructTypeInfo structTypeInfo = (StructTypeInfo) TypeConverter.convert(expr.getType());
+    return structTypeInfo.getAllStructFieldTypeInfos();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected List<String> getStructNameList(RexNode expr) {
+    StructTypeInfo structTypeInfo = (StructTypeInfo) TypeConverter.convert(expr.getType());
+    return structTypeInfo.getAllStructFieldNames();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected boolean isORFuncCallExpr(RexNode expr) {
+    return expr.isA(SqlKind.OR);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected boolean isANDFuncCallExpr(RexNode expr) {
+    return expr.isA(SqlKind.AND);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected boolean isPOSITIVEFuncCallExpr(RexNode expr) {
+    return expr.isA(SqlKind.PLUS_PREFIX);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected boolean isNEGATIVEFuncCallExpr(RexNode expr) {
+    return expr.isA(SqlKind.MINUS_PREFIX);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexNode setTypeInfo(RexNode expr, TypeInfo type) throws CalciteSemanticException {
+    RelDataType t = TypeConverter.convert(type, rexBuilder.getTypeFactory());
+    if (expr instanceof RexCall) {
+      RexCall call = (RexCall) expr;
+      return rexBuilder.makeCall(t,
+          call.getOperator(), call.getOperands());
+    } else if (expr instanceof RexInputRef) {
+      RexInputRef inputRef = (RexInputRef) expr;
+      return rexBuilder.makeInputRef(t, inputRef.getIndex());
+    } else if (expr instanceof RexLiteral) {
+      RexLiteral literal = (RexLiteral) expr;
+      return rexBuilder.makeLiteral(RexLiteral.value(literal), t, false);
+    }
+    throw new RuntimeException("Unsupported expression type: " + expr.getClass().getCanonicalName());
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected boolean canConvertCASEIntoCOALESCEFuncCallExpr(GenericUDF genericUDF, List<RexNode> inputs) {
```

Comment:
I don't really like this method...Ibelieve this functionality should live somewhere else instead (HivePointLookupOptimizer) ?

Reply:
I am not sure... The method is utterly ugly because it mixes UDF with Calcite RexLiteral. That logic should simply not be in `TypeCheckProcFactory`. The reason why I had to leave it there is because we convert back CASE into COALESCE while going back from Calcite to Hive. However, I think from AST to Calcite this is not even needed. Hence, I am just going to return `false` for the `RexNodeExprFactory` implementation and cross my fingers...

Reply:
In the future, I hope we can get rid of that method all together.

Reply:
https://issues.apache.org/jira/browse/HIVE-23226